launcher.py

```diff
@@ -12,94 +12,149 @@
+        logging.config.dictConfig(Config.DEFAULT_LOGGING_CONFIG)
 
-    if process.returncode != 0:
-        raise subprocess.CalledProcessError(process.returncode, command)
 
+class Utils:
+    VENV_NAME = "virtual_env"  # For downloading python packages
+    VENV_DIR_PATH = Path.cwd() / VENV_NAME  # Generate the path for the virtual environment
 
-def check_internet_connectivity(system: str) -> bool:
-    """
-    Check if the system has internet connectivity by pinging Google's public DNS server.
-    """
-    try:
-        if system == "Windows":
-            cmd = "ping -n 1 8.8.8.8"
-        else:
-            cmd = "ping -c 1 8.8.8.8"
-        logging.debug("Checking internet connectivity... ")
-        run_command(cmd)
-        logging.debug("Internet connectivity OK")
-        return True
+    @staticmethod
+    def run_command(command: str, cwd: Optional[Path] = None):
+        """
+        Executes a shell command.
 
-    except Exception as e:
-        logging.error("No internet connection detected.")
-        logging.debug(f"Error: {e}")
-        return False
+        Args:
+            command (str): The command to run.
+            cwd (str, optional): The working directory. Defaults to None.
+
+        Raises:
+            subprocess.CalledProcessError: If any command fails.
+        """
+        logging.debug(f"Executing command: {command}")
+        process = subprocess.Popen(
+            command.split(),
+            cwd=cwd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+        stdout, stderr = process.communicate()  # wait for process to terminate
+        if stdout:
+            logging.debug(f"Command output: {stdout}")
+        if stderr:
+            logging.error(f"Command error: {stderr}")
+
+        if process.returncode != 0:
+            raise subprocess.CalledProcessError(process.returncode, command)
+
+    @staticmethod
+    def check_internet_connectivity(system) -> bool:
+        try:
+            cmd = "ping -n 1 8.8.8.8" if system == "Windows" else "ping -c 1 8.8.8.8"
+            Utils.run_command(cmd)
+            logging.debug("Internet connectivity OK")
+            return True
+        except Exception as e:
+            logging.error("No internet connection detected.")
+            logging.debug(f"Error: {e}")
+            return False
+
+    @staticmethod
+    def ensure_virtualenv_exists():
+        if Utils.VENV_DIR_PATH.exists():
+            logging.debug("Virtual environment already exists.")
+            return
+        logging.info(f"Generating virtual environment at: {Utils.VENV_DIR_PATH}")
+        Utils.run_command(f"{sys.executable} -m venv {str(Utils.VENV_DIR_PATH)} --system-site-packages")
+        logging.info("Successfully generated Virtual environment.")
+
+    @staticmethod
+    def get_virtualenv_folder() -> Path:
+        """
+        get the path of the virtual environment binaries folder
+        """
```

Comment:
please stick to the convention of `directory` instead of the `folder, but in this case, this should be named `get_venv_path()` as you clearly say in the docstring, it returns the path and not a folder.

Reply:
Change name to `get_venv_binaries_directory`?

Reply:
What I meant was to prefer directory over folder in general, but in this case we're getting a path so we should say this in the method name. So, it'd be `get_venv_path()`.

Reply:
Yeah! I get that point. :+1:
what i meant actually was venv directory is virtual_env and with this function we are getting the directory in which the binaries reside so we should name it get_venv_binaries_directory?

Reply:
Ah, okay I get it now. In that case I'd still prefer `get_venv_binaries_path()` as it's returning a Path object.